stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -1295,12 +1896,39 @@ class EnumTypeInfoBuilder {
         Alignment = FixedDescriptor->getAlignment();
         NumExtraInhabitants = FixedDescriptor->NumExtraInhabitants;
         BitwiseTakable = FixedDescriptor->isBitwiseTakable();
+        unsigned Stride = ((Size + Alignment - 1) & ~(Alignment - 1));
+        if (Stride == 0)
+          Stride = 1;
+
+        auto PayloadSize = EnumTypeInfo::getPayloadSizeForCases(Cases);
+        BitMask spareBitsMask(PayloadSize);
+        auto validSpareBitsMask = populateSpareBitsMask(Cases, spareBitsMask);
+
+        if (!validSpareBitsMask) {
+          // If we couldn't correctly determine the spare bits mask,
+          // return a TI that will always fail when asked for XIs or value.
+          return TC.makeTypeInfo<UnsupportedEnumTypeInfo>(
+            Size, Alignment, Stride, NumExtraInhabitants,
+            BitwiseTakable, EnumKind::MultiPayloadEnum, Cases);
+        } else if (spareBitsMask.isZero()) {
+          // Simple case that does not use spare bits
+          // This is correct as long as our local spare bits calculation
+          // above only returns an empty mask when the mask is really empty,
+          return TC.makeTypeInfo<SimpleMultiPayloadEnumTypeInfo>(
+            Size, Alignment, Stride, NumExtraInhabitants,
+            BitwiseTakable, Cases);
+        } else {
+          // General case can mix spare bits and extra discriminator
+          // It obviously relies on having an accurate spare bit mask.
+          return TC.makeTypeInfo<MultiPayloadEnumTypeInfo>(
+            Size, Alignment, Stride, NumExtraInhabitants,
+            BitwiseTakable, Cases, spareBitsMask);
+        }
       } else {
-        // Dynamic multi-payload enums always use an extra tag to differentiate
-        // between cases
+        // Dynamic multi-payload enums cannot use spare bits, so they
+        // always use a separate tag value:
```

Comment:
@jckarter @slavapestov -- Is my logic here correct:  If there's not a FixedDescriptor (according to the logic above), then this MPE does not use spare bits, so the `SimpleMultipayloadEnumTypeInfo` can handle it correctly.

Reply:
Yeah, that should be the case that if the enum ever needs dynamic type layout, we don't use spare bits.

Reply:
Thanks for confirming that.  I've just added a commit that keeps this case but diverts all fixed-layout MPEs to the `Unsupported` bucket.